KAFKA-17236: handle local log deletion when remote.copy.disabled=true

showuon opened the pull request:
Handle local log deletion when `remote.copy.disabled=true`. 
In this PR, I did:
1. added `hasRemoteLogs` method to check if `remote.log.copy.disabled=true` && `log start offset != local log start offset` before checking local logs and update log start offset
2. Changed to use `retention.bytes`/`retention.ms` when remote copy disabled and remote logs are empty.
3. added tests

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

m1a2st committed:
KAFKA-16666 Migrate OffsetMessageFormatter to tools module (#16689)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

Commit:
KAFKA-17188: Ensure login and callback handler are closed upon encountering LoginException (#16724)

Reviewers: TengYao Chi <kitingiao@gmail.com>, Rajini Sivaram <rajinisivaram@googlemail.com>

C0urante committed:
MINOR: Unify Log4j levels across Connect runtime and MirrorMaker 2 integration tests (#16733)

Reviewers: Yash Mayya <yash.mayya@gmail.com>

linu-shibu committed:
KAFKA-16899 MembershipManagerImpl: rebalanceTimeoutMs variable name changed to commitTimeoutDuringReconciliation (#16334)

As mentioned in the ticket, the config property name "rebalanceTimeoutMs" is misleading since the property only handles the client's commit portion of the process.It is used in MembershipManagerImpl as a means to limit the client's efforts in the case where it is repeatedly trying to commit but failing. Considering the same, the property name has been updated to "commitTimeoutDuringReconciliation" (suggested in ticket) in GroupRebalanceConfig and in all other classes where the property/variable is used

Reviewers: Kirk True <kirk@kirktrue.pro>, Chia-Ping Tsai <chia7712@gmail.com>

mjsax committed:
MINOR: update EosIntegrationTest (#16697)

Refactor test to move off deprecated `transform()` in favor of
`process()`.

Reviewers: Bill Bejeck <bill@confluent.io>

mjsax committed:
MINOR: update EosV2UpgradeIntegrationTest (#16698)

Refactor test to move off deprecated `transform()` in favor of
`process()`.

Reviewers: Bill Bejeck <bill@confluent.io>

mjsax committed:
MINOR: update KStreamAggregationIntegrationTest (#16699)

Refactor test to move off deprecated `transform()` in favor of
`process()`.

Reviewers: Bill Bejeck <bill@confluent.io>

mjsax committed:
MINOR: update StandbyTaskCreationIntegrationTest (#16700)

Refactor test to move off deprecated `transform()` in favor of
`process()`.

Reviewers: Bill Bejeck <bill@confluent.io>

mjsax committed:
MINOR: update SuppressionDurabilityIntegrationTest (#16740)

Refactor test to move off deprecated `transform()` in favor of `process()`.

Reviewers: Bill Bejeck <bill@confluent.io>

loicgreffier committed:
KAFKA-16448: Unify class cast exception handling for both key and value (#16736)

Part of KIP-1033. Minor code cleanup.

Reviewers: Matthias J. Sax <matthias@confluent.io>

k0b3rIT committed:
KAFKA-17192 Fix MirrorMaker2 worker config does not pass config.provi… (#16678)

Reviewers: Chris Egerton <chrise@aiven.io>

sebastienviale committed:
KAFKA-16448: Handle processing exceptions in punctuate (#16300)

This PR is part of KIP-1033 which aims to bring a ProcessingExceptionHandler to Kafka Streams in order to deal with exceptions that occur during processing.

This PR actually catches processing exceptions from punctuate.

Co-authored-by: Dabz <d.gasparina@gmail.com>
Co-authored-by: loicgreffier <loic.greffier@michelin.com>

Reviewers: Bruno Cadonna <bruno@confluent.io>, Matthias J. Sax <matthias@confluent.io>

mjsax committed:
MINOR: update StreamsGraphTest (#16741)

Refactor test to move off deprecated `transform()` in favor of `process()`.

Reviewers: Bill Bejeck <bill@confluent.io>

showuon commented:
@kamalcph, if you have time, could you help have an early review for this PR? It depends on https://github.com/apache/kafka/pull/16653 , so I can change the target to `trunk` after it is merged. Thank you.

kamalcph commented:
Overall LGTM!

We may have to update the logic of `localRetentionSize` and `localRetentionMs` methods when remoteCopy gets disabled. Once all the remote log segments are deleted, then we have to use the `retention.bytes` and `retention.ms` for local log deletion instead of `local.retention.bytes` and `local.retention.ms`. 

Once `remote.copy.disabled` is set to true, whether to use local-log retention / complete-log retention for local-log segment deletion is subjective. 

IMO, we can add an validation to inform the user to set the local and complete retention to the same values when `remote.copy.disabled` is set to true. The user can take an informed decision when disabling the remote copy. Assume that the user configured the total retention for the topic to be 1 TB / 1 year and local retention to be 10 GB / 1 day. If we disable the remote copy, then the local logs are not eligible for deletion until all the remote log gets deleted. So, the local disk size will start to grow to 1 TB / 1 year and can lead to disk getting full in the broker.

m1a2st committed:
MINOR: Remove GroupMetadataManager#isGroupOpenForProducer unused method (#16731)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

sjhajharia committed:
MINOR: Cleanup storage module (#16207)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

cmccabe committed:
KAFKA-17234: Partition.tryCompleteDelayedRequests should not throw (#16760)

`Partition.tryCompleteDelayedRequests` should not throw since the
contexts it is called from are not expecting it, and the exceptions are
not related to those contexts.

Reviewers: Justine Olshan <jolshan@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

apoorvmittal10 committed:
KAFKA-17230: Fix for consumer node latency metrics (#16755)

Kafka Consumer client registers node/connection latency metrics in Selector.java but the values against the metric is never recorded. This seems to be an issue since inception. 

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Andrew Schofield <aschofield@confluent.io>, Jun Rao <junrao@gmail.com>

ahuang98 committed:
KAFKA-16521; kafka-metadata-quorum describe command changes for KIP-853 (#16759)

describe --status now includes directory id and endpoint information for voter and observers.
describe --replication now includes directory id.

Reviewers: Colin P. McCabe <cmccabe@apache.org>, José Armando García Sancio <jsancio@apache.org>

C0urante committed:
MINOR: Add extended cluster readiness check for Connect's embedded Kafka cluster (#16757)

Reviewers: Greg Harris <greg.harris@aiven.io>

m1a2st committed:
KAFKA-16666 Migrate GroupMetadataMessageFormatter` to tools module (#16748)

we need to migate GroupMetadataMessageFormatter from scala code to java code,and make the message format is json pattern

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

showuon commented:
> Overall LGTM!
> 
> We may have to update the logic of `localRetentionSize` and `localRetentionMs` methods when remoteCopy gets disabled. Once all the remote log segments are deleted, then we have to use the `retention.bytes` and `retention.ms` for local log deletion instead of `local.retention.bytes` and `local.retention.ms`.
> 
> Once `remote.copy.disabled` is set to true, whether to use local-log retention / complete-log retention for local-log segment deletion is subjective.
> 
> IMO, we can add an validation to inform the user to set the local and complete retention to the same values when `remote.copy.disabled` is set to true. The user can take an informed decision when disabling the remote copy. Assume that the user configured the total retention for the topic to be 1 TB / 1 year and local retention to be 10 GB / 1 day. If we disable the remote copy, then the local logs are not eligible for deletion until all the remote log gets deleted. So, the local disk size will start to grow to 1 TB / 1 year and can lead to disk getting full in the broker.

Good point! I've updated to use `retention.bytes` and `retention.ms` when remote copy disabled and remote logs are empty. Also added tests. Thanks for pointing it out!

kamalcph commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -1464,6 +1464,21 @@ class UnifiedLog(@volatile var logStartOffset: Long,
+   * @return true if this partition has (or will have) remote logs
+   */
+  private def hasRemoteLogs(): Boolean = {
+    remoteLogEnabled() && (!config.remoteCopyDisabled() || logStartOffset != localLogStartOffset())
```
`logStartOffset` may not converge with `localLogStartOffset` if there is a leader change post disabling the remote copy since the segment roll-over are independent of the replicas.

showuon replied:
hmm... any better idea to check it?

showuon replied:
`logStartOffset` >= `localLogStartOffset` ?

showuon replied:
`logStartOffset` >= `highestOffsetInRemoteStorage`?

kamalcph replied:
Sorry for the confusion. The existing check itself seems sufficient since we always move the LLSO to be the max (LSO, LLSO). I went over multiple paths and it looks good. 

https://sourcegraph.com/github.com/apache/kafka@trunk/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L173

showuon replied:
Awesome! Thank you!

sebastienviale committed:
KAFKA-16448: Update documentation (#16776)

Updated docs for KIP-1033.

Reviewers: Matthias J. Sax <matthias@confluent.io>

C0urante committed:
MINOR: Fix consumer group warmup in MirroMaker 2 integration tests (#16771)

Reviewers: Greg Harris <greg.harris@aiven.io>

mjsax committed:
KAFKA-9738: Deprecate old Processor API (#16742)

Implements KIP-1070, which deprecates:
 - Transformer
 - TransformerSupplier
 - ValueTransformer
 - ValueTransformerSupplier
 - MockProcessorContext

Reviewers: Bill Bejeck <bill@confluent.io>

C0urante committed:
MINOR: Fix ConnectWorkerIntegrationTest::testRequestTimeouts (#16772)

Reviewers: Greg Harris <greg.harris@aiven.io>

cmccabe committed:
KAFKA-16518: Implement KIP-853 flags for storage-tool.sh (#16669)

As part of KIP-853, storage-tool.sh now has two new flags: --standalone, and --initial-voters. This PR implements these two flags in storage-tool.sh.

There are currently two valid ways to format a cluster:

The pre-KIP-853 way, where you use a statically configured controller quorum. In this case, neither --standalone nor --initial-voters may be specified, and kraft.version must be set to 0.

The KIP-853 way, where one of --standalone and --initial-voters must be specified with the initial value of the dynamic controller quorum. In this case, kraft.version must be set to 1.

This PR moves the formatting logic out of StorageTool.scala and into Formatter.java. The tool file was never intended to get so huge, or to implement complex logic like generating metadata records. Those things should be done by code in the metadata or raft gradle modules. This is also useful for junit tests, which often need to do formatting. (The 'info' and 'random-uuid' commands remain in StorageTool.scala, for now.)

Reviewers: José Armando García Sancio <jsancio@apache.org>

frankvicky committed:
KAFKA-16390: add `group.coordinator.rebalance.protocols=classic,consumer` to broker configs when system tests need the new coordinator (#16715)

Fix an issue that cause system test failing when using AsyncKafkaConsumer.
A configuration option, group.coordinator.rebalance.protocols, was introduced to specify the rebalance protocols used by the group coordinator. By default, the rebalance protocol is set to classic. When the new group coordinator is enabled, the rebalance protocols are set to classic,consumer.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, David Jacot <djacot@confluent.io>, Lianet Magrans <lianetmr@gmail.com>, Kirk True <kirk@kirktrue.pro>, Justine Olshan <jolshan@confluent.io>

FrankYang0529 committed:
KAFKA-16480: ListOffsets change should have an associated API/IBP version update (#16781)

   1. Use oldestAllowedVersion as 9 if using ListOffsetsRequest#EARLIEST_LOCAL_TIMESTAMP or ListOffsetsRequest#LATEST_TIERED_TIMESTAMP.
   2. Add test cases to ListOffsetsRequestTest#testListOffsetsRequestOldestVersion to make sure requireTieredStorageTimestamp return 9 as minVersion.
   3. Add EarliestLocalSpec and LatestTierSpec to OffsetSpec.
   4. Add more cases to KafkaAdminClient#getOffsetFromSpec.
   5. Add testListOffsetsEarliestLocalSpecMinVersion and testListOffsetsLatestTierSpecSpecMinVersion to KafkaAdminClientTest to make sure request builder has oldestAllowedVersion as 9.

Signed-off-by: PoAn Yang <payang@apache.org>

Reviewers: Luke Chen <showuon@gmail.com>

showuon committed:
KAFKA-16855: remote log disable policy in KRaft (#16653)

Reviewers: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>, Christo Lolov <lolovc@amazon.com>

frankvicky committed:
KAFKA-17245: Revert TopicRecord changes. (#16780)

Revert KAFKA-16257 changes because KIP-950 doesn't need it anymore.

Reviewers: Luke Chen <showuon@gmail.com>

lianetm committed:
MINOR: Fix consumer log on fatal error & improve memberId logging (#16720)

Fix log on consumer fatal error, to show member ID only if present. If no member ID the log will clearly indicate that the member has no member ID (instead of showing empty as it used to)

Apply same fix consistently to all other log lines that include member ID.

Reviewers: Kirk True <kirk@kirktrue.pro>, PoAn Yang <payang@apache.org>, TengYao Chi <kitingiao@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

TaiJuWu committed:
MINOR: add helper method to `ClusterInstance` to wait topic deletion (#16627)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

mjsax committed:
KAFKA-16448: Unify error-callback exception handling (#16745)

Follow up code cleanup for KIP-1033.

This PR unifies the handling of both error cases for exception handlers:
 - handler throws an exception
 - handler returns null

The unification happens for all 5 handler cases:
 - deserialzation
 - production / serialization
 - production / send
 - processing
 - punctuation

Reviewers:  Sebastien Viale <sebastien.viale@michelin.com>, Loic Greffier <loic.greffier@michelin.com>, Bill Bejeck <bill@confluent.io>

mumrah committed:
KAFKA-17146 Include note to remove migration znode (#16770)

When reverting the ZK migration, we must also remove the /migration ZNode in order to allow the migration to be re-attempted in the future.

Reviewers: Colin P. McCabe <cmccabe@apache.org>, Chia-Ping Tsai <chia7712@gmail.com>

brandboat committed:
KAFKA-16154: Broker returns offset for LATEST_TIERED_TIMESTAMP (#16783)

This pr support EarliestLocalSpec LatestTierSpec in GetOffsetShell, and add integration tests.

Reviewers: Luke Chen <showuon@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>, PoAn Yang <payang@apache.org>

brenden20 committed:
MINOR: Remove ConsumerTestBuilder.java (#16691)

The purpose of this PR is to remove ConsumerTestBuilder.java since it is no longer needed. The following PRs have eliminated the use of ConsumerTestBuilder:
#14930
#16140
#16200
#16312

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

showuon committed:
KAFKA-17236: handle remote copy disabled on local log deletion

showuon committed:
KAFKA-17236: rename to `remote.log.copy.disable`

showuon committed:
KAFKA-17236: fix the test comments

showuon committed:
KAFKA-17236: adopt retention.ms/bytes when remote.log.copy.disable=true

showuon committed:
KAFKA-17236: add retention validation when remote log copy disabled
